drivers/gpio/gpio_pca95xx.c

```diff
@@ -0,0 +1,675 @@
+#define REG_OUTPUT_PORT_CONF		0x4F
+
+/* Driver flags */
+#define PCA_HAS_PUD			BIT(0)
+
+/** Configuration data */
+struct gpio_pca95xx_config {
+	/** The master I2C device's name */
+	const char * const i2c_master_dev_name;
+
+	/** The slave address of the chip */
+	u16_t i2c_slave_addr;
+
+	u8_t capabilities;
+};
+
+/** Store the port 0/1 data for each register pair. */
+union gpio_pca95xx_port_data {
+	u16_t all;
+	u8_t port[2];
+	u8_t byte[2];
+};
+
+/** Runtime driver data */
+struct gpio_pca95xx_drv_data {
+	/* gpio_driver_data needs to be first */
+	struct gpio_driver_data common;
+
+	/** Master I2C device */
+	struct device *i2c_master;
+
+	struct {
+		union gpio_pca95xx_port_data output;
+		union gpio_pca95xx_port_data dir;
+		union gpio_pca95xx_port_data pud_en;
+		union gpio_pca95xx_port_data pud_sel;
+	} reg_cache;
+
+	struct k_sem lock;
+};
+
+/**
+ * @brief Read both port 0 and port 1 registers of certain register function.
+ *
+ * Given the register in reg, read the pair of port 0 and port 1.
+ *
+ * @param dev Device struct of the PCA95XX.
+ * @param reg Register to read (the PORT0 of the pair of registers).
+ * @param buf Buffer to read data into.
+ *
+ * @return 0 if successful, failed otherwise.
+ */
+static int read_port_regs(struct device *dev, u8_t reg,
+			   union gpio_pca95xx_port_data *buf)
+{
+	const struct gpio_pca95xx_config * const config =
+		dev->config->config_info;
+	struct gpio_pca95xx_drv_data * const drv_data =
+		(struct gpio_pca95xx_drv_data * const)dev->driver_data;
+	struct device * const i2c_master = drv_data->i2c_master;
+	u16_t i2c_addr = config->i2c_slave_addr;
+	int ret;
+
+	ret = i2c_burst_read(i2c_master, i2c_addr, reg, buf->byte, 2);
+	if (ret) {
+		LOG_ERR("PCA95XX[0x%X]: error reading register 0x%X (%d)",
+			i2c_addr, reg, ret);
+		goto error;
+	}
+
+	LOG_DBG("PCA95XX[0x%X]: Read: REG[0x%X] = 0x%X, REG[0x%X] = 0x%X",
+		i2c_addr, reg, buf->byte[0], (reg + 1), buf->byte[1]);
+
+error:
+	return ret;
+}
+
+/**
+ * @brief Write both port 0 and port 1 registers of certain register function.
+ *
+ * Given the register in reg, write the pair of port 0 and port 1.
+ *
+ * @param dev Device struct of the PCA95XX.
+ * @param reg Register to write into (the PORT0 of the pair of registers).
+ * @param buf Buffer to write data from.
+ *
+ * @return 0 if successful, failed otherwise.
+ */
+static int write_port_regs(struct device *dev, u8_t reg,
+			    union gpio_pca95xx_port_data *buf)
+{
+	const struct gpio_pca95xx_config * const config =
+		dev->config->config_info;
+	struct gpio_pca95xx_drv_data * const drv_data =
+		(struct gpio_pca95xx_drv_data * const)dev->driver_data;
+	struct device * const i2c_master = drv_data->i2c_master;
+	u16_t i2c_addr = config->i2c_slave_addr;
+	int ret;
+
+	LOG_DBG("PCA95XX[0x%X]: Write: REG[0x%X] = 0x%X, REG[0x%X] = "
+		"0x%X", i2c_addr, reg, buf->byte[0], (reg + 1),
+		buf->byte[1]);
+
+	ret = i2c_burst_write(i2c_master, i2c_addr, reg, buf->byte, 2);
```

Comment:
We need to be careful here since on a big endian MCU the value in `buf->byte` will have LSB located in `byte[1]`. Check implementation of `i2c_reg_write_word_be` function in `gpio_sx1509b.c` driver. The difference is that here the value needs to be little endian (if I read PCAL9535A datasheet correctly). The same holds true for the `read_port_regs` function.